Delay loading JNI refs

tajila opened the pull request:
Delay loading JNI refs

The openj9.criu module is loaded after java.base so any operations that
require InternalCRIUSupport early on in JVM startup will hit an
assertion failure when attempting to setup the CRIU JNI refs.

Fixes: https://github.com/eclipse-openj9/openj9/issues/14821

Signed-off-by: Tobi Ajila <atobia@ca.ibm.com>

tajila commented:
Jenkins test sanity xlinuxcriu jdk11

tajila commented:
Jenkins test sanity win jdk8

DanHeidinga commented on runtime/criusupport/criusupport.cpp:
```diff
@@ -94,7 +94,6 @@ Java_openj9_internal_criu_InternalCRIUSupport_isCRIUSupportEnabledImpl(JNIEnv *e
 		}
 #endif /* defined(LINUX) */
 	}
-	setupJNIFieldIDs(env);
```
I thought we did this early so that by the time checkpoint() was called, we'd be sure it would work (at least from an infrastructure perspective).

Would only setting up the JNI fields if `res == JNI_TRUE` avoid the assertion?

```c
if (res == JNI_TRUE) {
    setupJNIFieldIDs(env);
}
```

tajila replied:
> Would only setting up the JNI fields if res == JNI_TRUE avoid the assertion?

No, we moved the `isCRIUSupportEnabledImpl` native to java.base because it is needed very early by the security providers. The classes being loaded in `setupJNIFieldIDs` are in openj9.criu module which is loaded later.

> I thought we did this early so that by the time checkpoint() was called, 

Initially we did, but more recently we have tried to de-tangle things the require libcriu.so to be laoded, and status checks like `isCRIUSupportEnabledImpl`

The `setupJNIFieldIDs` call cannot fail once openj9.criu module is loaded so we are free to move it

tajila commented:
Ive noticed some other issues, I will make an update

tajila commented:
I've updated the PR.

I removed `criu_init_opts` in `isCRIUSupportEnabledImpl` because its a redundant check, we do it again in checkpointImpl.

Also, as stated above, I would like to be able to query if the JVM enables CRIU without loading the library. I would like to only load the library in CRIUSupport static init so if there is an error in loading the library, it is captured in an unstatisfiedLinkageError rather than hard failures with `ld`.

tajila commented:
Jenkins compile xlinuxcriu jdk11
